File: week2/extraction.py
import logging
import time

from langchain_core.exceptions import OutputParserException
from langchain_core.output_parsers import PydanticOutputParser
from pydantic import ValidationError

import config
import models
import prompts
import schemas

logger = logging.getLogger("week2.extraction")

parser = PydanticOutputParser(pydantic_object=schemas.SupportTicket)


async def extract_support_ticket(email_text, request_id="-", max_retries=None):
    """Reject-and-retry loop for Exercise 2. Each attempt: call the model in JSON mode,
    then run the raw text through PydanticOutputParser, which both parses the JSON and
    validates it against SupportTicket's field types/enums. A malformed reply (bad JSON,
    a priority value outside the allowed enum, a badly-formed email address, ...) raises
    instead of silently passing through, and we retry rather than return garbage."""
    max_retries = max_retries or config.EXTRACTION_MAX_RETRIES
    chat = models.get_json_mode_chat_model(config.EXTRACTION_MODEL)
    messages = prompts.extraction_prompt_template.format_messages(
        format_instructions=parser.get_format_instructions(),
        email_text=email_text,
    )

    last_error = None
    for attempt in range(1, max_retries + 1):
        start = time.perf_counter()
        try:
            response = await models.ainvoke_with_retry(chat, messages)
            ticket = parser.parse(response.content)
            elapsed = round(time.perf_counter() - start, 2)
            usage = models.extract_token_usage(response)
            logger.info(
                f"[{request_id}] extraction attempt {attempt}/{max_retries} succeeded in {elapsed}s "
                f"(tokens={usage})"
            )
            return {"ticket": ticket.model_dump(), "attempts": attempt, "error": None}
        except (OutputParserException, ValidationError) as error:
            last_error = f"malformed model output: {error}"
            logger.warning(
                f"[{request_id}] extraction attempt {attempt}/{max_retries} rejected - {str(error)[:200]}"
            )
        except Exception as error:
            last_error = str(error)
            logger.error(f"[{request_id}] extraction attempt {attempt}/{max_retries} failed - {last_error[:200]}")

    return {
        "ticket": None,
        "attempts": max_retries,
        "error": f"could not extract a valid ticket after {max_retries} attempts. Last error: {last_error}",
    }

File: week2-js/src/extraction.ts
import { SupportTicketSchema, type SupportTicket } from "./schemas.js";
import * as config from "./config.js";
import * as models from "./models.js";
import { extractionPromptTemplate } from "./prompts.js";

// Hand-written JSON-shape instructions, playing the same role as Python's
// PydanticOutputParser.get_format_instructions() (Level 1 of structured output:
// tell the model the exact shape you expect, in the prompt itself).
function getFormatInstructions(): string {
  return (
    "Respond with a single JSON object matching exactly this shape (no extra text, no markdown fences):\n" +
    "{\n" +
    '  "customer_name": string,        // full name, or "Unknown" if not stated\n' +
    '  "customer_email": string,       // the email address exactly as written in the message\n' +
    '  "priority": "low" | "medium" | "high" | "urgent",\n' +
    '  "issue_type": "billing" | "technical" | "account" | "feature_request" | "bug" | "other",\n' +
    '  "summary": string                // one to two sentence summary of the issue\n' +
    "}"
  );
}

export interface ExtractionResult {
  ticket: SupportTicket | null;
  attempts: number;
  error: string | null;
}

/** Reject-and-retry loop for Exercise 2. Each attempt: call the model in JSON mode,
 * then JSON.parse the raw text (syntax check) and run it through
 * SupportTicketSchema.parse (shape/enum check - the zod equivalent of Pydantic
 * validation). A malformed reply (bad JSON, a priority value outside the allowed
 * enum, a badly-formed email address, ...) throws instead of silently passing
 * through, and we retry rather than return garbage. */
export async function extractSupportTicket(
  emailText: string,
  requestId = "-",
  maxRetries: number = config.EXTRACTION_MAX_RETRIES
): Promise<ExtractionResult> {
  const chat = models.getJsonModeChatModel(config.EXTRACTION_MODEL);
  const messages = await extractionPromptTemplate.formatMessages({
    format_instructions: getFormatInstructions(),
    email_text: emailText,
  });

  let lastError: string | null = null;

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    const start = performance.now();
    try {
      const response = await models.ainvokeWithRetry(chat, messages);
      const raw = JSON.parse(response.content as string); // throws SyntaxError on bad JSON
      const ticket = SupportTicketSchema.parse(raw); // throws ZodError on bad shape/enum
      const elapsed = Math.round((performance.now() - start) / 10) / 100;
      const usage = models.extractTokenUsage(response);
      console.log(
        `[${requestId}] extraction attempt ${attempt}/${maxRetries} succeeded in ${elapsed}s (tokens=${JSON.stringify(usage)})`
      );
      return { ticket, attempts: attempt, error: null };
    } catch (error) {
      const isMalformed = error instanceof SyntaxError || (error as { name?: string })?.name === "ZodError";
      if (isMalformed) {
        lastError = `malformed model output: ${error}`;
        console.warn(`[${requestId}] extraction attempt ${attempt}/${maxRetries} rejected - ${String(error).slice(0, 200)}`);
      } else {
        lastError = String(error);
        console.error(`[${requestId}] extraction attempt ${attempt}/${maxRetries} failed - ${lastError.slice(0, 200)}`);
      }
    }
  }

  return {
    ticket: null,
    attempts: maxRetries,
    error: `could not extract a valid ticket after ${maxRetries} attempts. Last error: ${lastError}`,
  };
}
